### PR TITLE
RFC Add BLAKE3 Checksum to ZFS

### PR DESCRIPTION
**Pull request by mcmilk**
Add [BLAKE3](https://github.com/BLAKE3-team/BLAKE3) to ZFS (done) and maybe do also some more micro benchmarking within ZFS (planned).

Please fill out the following template, which will help other contributors review your Pull Request.
@behlendorf, @ahrens and @allanjude 

This PR integrates Blake3 as new checksum algorithm. It will be faster then edon-r but provide state-of-the-art cryptographic hashing for dedup and so on. I created some first [benchmarks](https://gist.github.com/mcmilk/f929b5c2c3383be301c907f5b9dfeb86) using this [script](https://github.com/mcmilk/zfs/blob/81be18af531f55f0ec5bfdd3443c24f13b3bb057/tmp/do-benchmark.sh) completly within memory using my Laptop.

### Motivation and Context

I think, the checksumming in ZFS is a very strong feature. This patch adds Blake3 - which is with block sizes > 1KiB a lot faster then all other currently integrated hashes. For data blocks from 256bytes to 128kb the performance is like the other ones.

### Description

Hello ZFS developer - this is just a first RFC for review before adding Blake3 checksum algorithm to zfs.
The real PR will be seperated into smaller pieces and will have some real work an the man-page stuff and testing stuff ...  ;-)

**This is an overview new files:**
new file:   include/sys/blake3.h
new file:   module/icp/algs/blake3/blake3.c
new file:   module/icp/algs/blake3/blake3_generic.c
new file:   module/icp/algs/blake3/blake3_impl.c
new file:   module/icp/algs/blake3/blake3_neon.c
new file:   module/icp/algs/blake3/blake3_x86-64.c
new file:   module/icp/asm-x86_64/blake3/blake3_avx2.S
new file:   module/icp/asm-x86_64/blake3/blake3_avx512.S
new file:   module/icp/asm-x86_64/blake3/blake3_sse2.S
new file:   module/icp/asm-x86_64/blake3/blake3_sse41.S
new file:   module/icp/include/blake3/blake3_impl.h
new file:   module/zfs/blake3_zfs.c

**These are changed:**
modified:   AUTHORS
modified:   include/sys/Makefile.am
modified:   include/sys/zio.h
modified:   include/sys/zio_checksum.h
modified:   include/zfeature_common.h
modified:   lib/libicp/Makefile.am
modified:   lib/libzpool/Makefile.am
modified:   man/man5/zpool-features.5
modified:   man/man8/zfsprops.8
modified:   module/icp/Makefile.in
modified:   module/zcommon/zfeature_common.c
modified:   module/zcommon/zfs_prop.c
modified:   module/zfs/Makefile.in
modified:   module/zfs/zio_checksum.c

### How Has This Been Tested?
Currently I only tested this on some VM's and the [do-benchmark.sh](https://gist.github.com/mcmilk/f929b5c2c3383be301c907f5b9dfeb86) script was run on my Dell Notebook (i7-3632QM, 2TB SSD, 16GB RAM). I will do more tests.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [x] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [x] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [x] I have run the ZFS Test Suite with this change applied.
- [ ] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

File: module/zfs/blake3_zfs.c [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+abd_checksum_blake3_native(abd_t *abd, uint64_t size, const void *ctx_template,
+    zio_cksum_t *zcp)
+{
+	/*
+	 * CC [M]  module/zfs/blake3_zfs.o
+	 * module/zfs/blake3_zfs.c: In function ‘abd_checksum_blake3_native’:
+	 * module/zfs/blake3_zfs.c:56:1: warning: the frame size of 1920 bytes
+	 *     is larger than 1024 bytes [-Wframe-larger-than=]
+	 */
+	BLAKE3_CTX ctx;
```
Comment (allanjude) [inline module/zfs/blake3_zfs.c:55]: Yeah, that large stack variable may be a problem, you might need to switch to kmem_cache_alloc or similar.

**Review by allanjude — COMMENTED** [review]
Things that we need to look at:
1) Make sure the feature flag is activated when the property is set (see what we did for zstd), otherwise it may cause similar issues if someone sets the property but doesn't write any new blocks, so the feature flag doesn't get activated)

2) How will that impact the send stream format? Does it need a send stream feature flag?
IIRC, the send stream format always uses a fletcher4 checksum? Does this apply to raw streams?


Otherwise, the integration into ZFS looks good to me, but I've not reviewed the blake3 implementation.

File: module/zfs/blake3_zfs.c (hunk 1 above)
Comment (mcmilk) [reply module/zfs/blake3_zfs.c:55]: Yes, I got the warning also - but not an all machines... thats why I forgot about it again ;)
I will try the switch to mem_cache_alloc() and will take a look, if it performs ...

Comment (mcmilk) [reply module/zfs/blake3_zfs.c]: - feature flag ist done
- How will that impact the send stream format? -> the Hash will be used like edonr or skein (mainly dedup)
- kmem_cache_alloc() works nicely

Comment (mcmilk) [reply module/zfs/blake3_zfs.c]: Hey @allanjude, I have no idea why the FreeBSD tests are failing in the moment. They tell that Blake3_Init(), Blake3_Update() and Final() can not be found. But the ipc.ko module does export them.

Is there a need to create module/icp/io/blake3_mod.c or sth. ?
In general - should I provide this module - is it needed ?

Thanks for your help.

**Comment by ericonr** [comment]
Maybe it would be best to have the vendoring in of BLAKE3 / setting up the build system as a separate commit from plumbing it into ZFS?

**Comment by mcmilk** [comment]
> Maybe it would be best to have the vendoring in of BLAKE3 / setting up the build system as a separate commit from plumbing it into ZFS?

I will split the big commit into 3-4 commits and seperate the things.
This is planned... currently I am using the tests here for checking that all things work properly. We have time for this ... ;-)

**Comment by mko-io** [comment]
There's a new [PR](https://github.com/BLAKE3-team/BLAKE3/pull/181) to add GPU Vulkan support, which offload the checksum even more off CPU

**Comment by rincebrain** [comment]
Hiya,
Any interest in refreshing this? I'd be curious to see how it performs compared to the other hash options in practice on a wider variety of hardware...

File: module/zfs/blake3_zfs.c (hunk 1 above)
Comment (allanjude) [reply module/zfs/blake3_zfs.c]: FreeBSD doesn't build icp by default. You'll need to add the .c files you want to the list in Makefile.BSD to compile them into the main kernel module (see what @rincebrain did recently for edon-r)

**Comment by mcmilk** [comment]
> Hiya, Any interest in refreshing this? I'd be curious to see how it performs compared to the other hash options in practice on a wider variety of hardware...
> 
> edit: I am amused to discover that, when I went to go apply the few ASM changes between 0.3.7 and 1.2.0 (for an UB case), you appear to have already applied that change...back in April...when they only ran into it and fixed it a few weeks ago, heh. Whoops.

Yes, of cause.... but my time is currently very limited.
I am hoping for some free time in december... then I could review the whole patch and split it up.

**Comment by mcmilk** [comment]
I have close this PR because the the force pushed changes didn't compile successfully - I have no idea why... it looks, like git fetch + hard reset has problems with it.

**Pull request closed without merging**